web-app/src/main/java/com/google/sps/data/QuadTree.java

```diff
@@ -56,4 +58,85 @@ public void printTree() {
     System.out.printf("%n");
   }
 
+  public void insert(PoliceReport report) {
+    Node leafNode = findLeaf(root, report);
+    leafNode.numReports += 1;
+    leafNode.reports.add(report);
+    // If max capacity has been exceeded, create child nodes
+    if (leafNode.numReports > reportCapacity && leafNode.depth < maxDepth) {
+      leafNode.children = reallocateReports(leafNode.reports, leafNode.coordinates, leafNode.depth);
+      leafNode.leaf = false;
+      leafNode.reports = null;
+    }
+  }
+
+  // Find a leaf to insert the report in
+  private Node findLeaf(Node currentNode, PoliceReport report) {
+    double reportLat = report.getLat();
+    double reportLng = report.getLng();
+
+    while (!currentNode.leaf) {
+      currentNode.numReports += 1;
+      for (Node child : currentNode.children) {
+        if (child.coordinates.inRectangle(reportLat, reportLng)) {
+          currentNode = child;
+          break;
+        }
+      }
+    }
+
+    return currentNode;
+  }
+
+  private Node[] reallocateReports(ArrayList<PoliceReport> reports, Rectangle coordinates, int depth) {
+    Node[] children = new Node[4];
+
+    // Create empty children
+    for (Direction direction : Direction.values()) {
```

Comment:
Is there a better way to use enums? This feels a little clunky. I tried associating a value with the enums but it didn't really fix anything and just made the enum constructor unnecessarily complex.

Reply:
Yeah it's a bit clunky because the coordinates don't accept the enums. See if you can just create the four children explicitly without the loop:

```
children[NW] = new Node(newCoordinates, new ArrayList<PoliceReport>(), newDepth);
children[NE] = new Node(newCoordinates, new ArrayList<PoliceReport>(), newDepth);
...
```